### PR TITLE
SALTO-1993: add config changes to zendesk

### PR DESCRIPTION
**Pull request by noamhammer**
_Add config changes to zendesk_

---

_Additional context for reviewer_

---
_Release Notes_: 
_None_

---
_User Notifications_: 
_None_

### CODE REVIEW

File: packages/adapter-components/src/elements/ducktype/transformer.ts [hunk 1]
```diff
@@ -299,13 +310,26 @@ export const getAllElements = async ({
...
-      ...args,
-    }),
+    typeElementGetter: async args => {
+      try {
+        return await getTypeAndInstances({
+          ...elementGenerationParams,
+          ...args,
+        })
+      } catch (e) {
+        if (isErrorTurnToConfigSuggestion?.(e) === true) {
```
Comment (alonstern) [inline packages/adapter-components/src/elements/ducktype/transformer.ts:324]: nit: `if (isErrorTurnToConfigSuggestion?.(e)) {`

File: packages/zendesk-support-adapter/src/adapter.ts [hunk 2]
```diff
@@ -160,6 +165,8 @@ export default class ZendeskAdapter implements AdapterOperations {
       typeDefaults: this.userConfig.apiDefinitions.typeDefaults,
       getElemIdFunc: this.getElemIdFunc,
       hideTypes: this.userConfig.fetch.hideTypes,
+      isErrorTurnToConfigSuggestion: error =>
+        error instanceof clientUtils.HTTPError && (error.response.status === 403),
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/adapter.ts:169]: Any chance we will get this error in other scenarios? e.g., if the user's credentials expired, and we will remove all the types from the configuration? Maybe we should be more explicit here?

File: packages/zendesk-support-adapter/src/config_change.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+import _ from 'lodash'
+import { collections } from '@salto-io/lowerdash'
+import { InstanceElement, ElemID } from '@salto-io/adapter-api'
+import { elements as elementsUtils } from '@salto-io/adapter-components'
+
+import { configType, FETCH_CONFIG } from './config'
+
+const { makeArray } = collections.array
+
+const MESSAGE_INTRO = 'Salto failed to fetch some items from zendesk. '
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/config_change.ts:25]: nit: remove extra space

File: packages/zendesk-support-adapter/src/config_change.ts [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+      configType,
+      {
+        ...currentConfig.value,
+        fetch: {
+          ...currentConfig.value[FETCH_CONFIG],
+          includeTypes: newIncludeTypes,
+        },
+      },
+    )],
+    message: [MESSAGE_INTRO, '', MESSAGE_SUMMARY].join('\n'),
```
Comment (alonstern) [inline packages/zendesk-support-adapter/src/config_change.ts:55]: use `os.EOL`

**Review by alonstern — APPROVED** [review]

File: packages/zendesk-support-adapter/src/adapter.ts (hunk 2 above)
Comment (noamhammer) [reply packages/zendesk-support-adapter/src/adapter.ts:169]: In that case, we get unauthorzied error (401) and not Forbidden (403), so I think we are good

File: packages/adapter-components/src/elements/ducktype/transformer.ts [hunk 5]
```diff
@@ -56,6 +56,15 @@ type Entries = {
   nestedTypes: ObjectType[]
 }
 
+export type ConfigChangeSuggestion = {
+  typeToExclude: string
+}
+
+export type FetchElements<T> = {
+  configChanges: ConfigChangeSuggestion[]
+  elements: T
```
Comment (netama) [inline packages/adapter-components/src/elements/ducktype/transformer.ts:65]: why is this templated?

Comment (netama) [reply packages/adapter-components/src/elements/ducktype/transformer.ts:65]: also, is there a reason to add this only on ducktype?

File: packages/zendesk-support-adapter/src/adapter.ts (hunk 2 above)
Comment (netama) [reply packages/zendesk-support-adapter/src/adapter.ts:169]: I also think we should be more explicit and only do this for limited types

**Pull request merged**